lib/config/providers/chat_stream_provider.dart

```diff
@@ -14,70 +16,178 @@ typedef GroupMessageSubscriber =
 
 class ChatStreamNotifier extends AutoDisposeFamilyStreamNotifier<List<MessageModel>, String> {
   final _logger = Logger('ChatStreamNotifier');
-
   final GroupMessageSubscriber _subscriber;
 
+  Map<String, ChatMessage> _messageMap = {};
+  List<MessageModel> _optimisticMessages = [];
+  StreamController<List<MessageModel>>? _controller;
+  StreamSubscription<MessageStreamItem>? _rustStreamSubscription;
+
+  bool _isComputingState = false;
+  bool _computeStatePending = false;
+
   ChatStreamNotifier({
     GroupMessageSubscriber subscriber = subscribeToGroupMessages,
   }) : _subscriber = subscriber;
 
   @override
-  Stream<List<MessageModel>> build(String groupId) async* {
+  Stream<List<MessageModel>> build(String groupId) {
     final activePubkey = ref.watch(activePubkeyProvider);
+
+    _controller?.close();
```

Comment:
This will lead to side effects. because of the following:
- build() may run multiple times (dependency changes, re-listens, provider refersh).
- closing a controller or cancling a subscription during build() can race with active listeners and callbacks.
- Resetting internal state on each rebuild can drop messages or optimistic updates and cause UI flicker and jank.

I recommend the following:
- Do cleanups only in ref.onDispose
- And you need to restart the rust sub on dependency change, do it in a controlled restart method (avoid closing the stream that you're returning)
```suggestion
    _controller?.close();
```